HTTP client: Preserve request method and body for `307 Temporary Redirect` and `308 Permanent Redirect`

dinooo13 opened the pull request:
Should close #409 
303 See Other should change the method to GET all other redirects should leave it unchanged
(Source: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#redirection_messages)

I think there are some specialties for 300 and 304 which I ignored for now, but I could look further into it if you wish.

SimonFrings commented:
Hey @dinooo13 , thanks for looking into this one :+1:

This is an interesting topic and the approach looks nice. It may not seem like it but I think this not only fixes some stuff but also breaks some (for example the new request doesn't contain a body, so when redirecting a POST there is no data). It fixes the problem addressed in #409 but I also think it overloads this PR to use this new rule on every 300 status code (except 303).

I also looked a bit deeper into the [HTTP specification](https://datatracker.ietf.org/doc/html/rfc2616). For the majority of the 300 status codes, it says that the request method *may* change when redirecting, which depends a bit on the use case and who you're talking to. Because of all of this, I think it's best that this PR only focuses on the 307 (and 308) for now and that we tackle the remaining status codes in a following one (if necessary).

Additionally, the specification says, that the method for 303 should always be a GET when redirecting. You already covered this in your current PR but I was thinking: What if the original method was HEAD. The HEAD and GET methods are very similar to each other, the only difference is that the HEAD method doesn't request the body. If I use a HEAD method now and this changes into a GET when redirecting, I'm receiving more data than I originally wanted.

I am curious what your thoughts on this are! Should we add some kind of functionality that HEAD will stay the same when receiving a 303 (which is not explicitly mentioned inside the RFC)?

dinooo13 commented:
Okay I will update to only handle 307/308, yes it's better to address these when they actually become problems (YAGNI again 😄 ). Actually I found this in the RFC regarding 303 HEAD requests: 

> Except for responses to a HEAD request, the representation of a 303
   response ought to contain a short hypertext note with a hyperlink to
   the same URI reference provided in the Location header field.

dinooo13 commented:
Updated this PR to only handle 303, 307 and 308. 303 will be converted to GET. 307 and 308 will preserve their methods and body. All others should have the same behavior as before.

SimonFrings commented:
@dinooo13 Looking good, but I still think the logic for 303 should use the default behavior (HEAD stays HEAD and GET stays GET).

The logic behind your code works great, I am not quite sure if we're introducing to much complexity by adding two new methods. I came up with a shorter version of your code, still the same logic:
```php
private function onResponseRedirect(ResponseInterface $response, RequestInterface $request, Deferred $deferred, ClientRequestState $state)
    {
        // resolve location relative to last request URI
        $location = Uri::resolve($request->getUri(), $response->getHeaderLine('Location'));

        $request = $this->makeRedirectRequest($request, $location, $response->getStatusCode());
        $this->progress('redirect', array($request));

        if ($state->numRequests >= $this->maxRedirects) {
            throw new \RuntimeException('Maximum number of redirects (' . $this->maxRedirects . ') exceeded');
        }

        return $this->next($request, $deferred, $state);
    }
```

The first code block shows the `onResponseRedirect()` method. I have added `$response->getStatusCode()` as a parameter to the `makeRedirectRequest()` function call. The status code is the only part we need, this way we don't have to pass the whole response.

```php
private function makeRedirectRequest(RequestInterface $request, UriInterface $location, Int $statusCode)
    {
        $originalHost = $request->getUri()->getHost();
        $request = $request->withoutHeader('Host');
         
        if ($statusCode !== 307 || $statusCode !== 308) {
            $request = $request
                ->withoutHeader('Content-Type')
                ->withoutHeader('Content-Length');
        }

        // Remove authorization if changing hostnames (but not if just changing ports or protocols).
        if ($location->getHost() !== $originalHost) {
            $request = $request->withoutHeader('Authorization');
        }

        $body = null;
        if ($statusCode === 307 || $statusCode === 308) {
            $method = $request->getMethod();
            $body = $request->getBody();
        } else {
            // naïve approach..
            $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
        }

        return new Request($method, $location, $request->getHeaders(), $body);
    }
```

The second block shows the `makeRedirectRequest()` method. As you can see I've added the status code as an additional argument to this function. The rest follows the same logic as your suggestion, just a little less code. The `if`'s will only check for the 307 and 308 status codes and add the original method and body to the new request. Every other 3xx status code uses the default logic which is already implemented.

The only thing I don't know how to handle yet are streaming bodies. The original request would send the whole body before the redirect response arrives. This means the new request can't contain the same body (data is already sent). 

What are your thoughts on this?

dinooo13 commented:
Nice yes I think this really simplifies it. I found this about how Chrome handles this with the fetch API:
> Restricted redirects
Some forms of HTTP redirect require the browser to resend the body of the request to another URL. To support this, the browser would have to buffer the contents of the stream, which sort-of defeats the point, so it doesn't do that.

https://web.dev/fetch-upload-streaming/#restricted-redirects

I think this sums it up pretty good, it just makes no sense to redirect a stream.

SimonFrings commented:
Seems like this is the way to go then :+1: 

I also found this inside the [Guzzle documentation](https://guzzle3.readthedocs.io/http-client/http-redirects.html#redirects-and-non-repeatable-streams), throwing an exception in this case could be the answer for this ticket too. They're also using some kind of repeatable streams, this could be something for a follow up PR maybe ^^

dinooo13 commented:
Hey I updated the PR with your suggestions and added an exception and a test. Please have a close look at the new test I'm not 100% sure if it's correct. I'm also curious what you think about adding an exception with basically only a default message and a name should we even bother then? I think a hint in the documentation on reactphp.org would also be good to have.

SimonFrings commented on src/Io/Transaction.php:
```diff
@@ -255,25 +256,36 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
+        $body = null;
+        if ($statusCode === 307 || $statusCode === 308) {
+            $method = $request->getMethod();
+            if ($request->getBody() instanceof ReadableStreamInterface) {throw new CantRedirectStreamException();}
```
Maybe it's enough to throw a `RuntimeException` at this point, then we wouldn't need a new Exception class for this. The message could be something like "Unable to redirect request with streaming body". I think this should be enough to prevent our streaming body problem.

SimonFrings commented on tests/Io/TransactionTest.php:
```diff
@@ -666,6 +667,111 @@ public function testSomeRequestHeadersShouldBeRemovedWhenRedirecting()
+        $transaction = new Transaction($sender, $loop);
+        $promise = $transaction->send($request);
+
+        try {
+            Block\await($promise, $loop);
+            $this->fail();
+        } catch (CantRedirectStreamException $exception) {
+            $this->assertEquals('Can\'t redirect stream', $exception->getMessage());
+        }
```
If we use the `RuntimeException` as I suggested above we could use the following code inside the new test:
```php
$this->setExpectedException('RuntimeException', 'Unable to redirect request with streaming body');
```
The `setExpectedException()` method is a work around I added to the project a while ago to ensure that the same logic works across our currently supported PHP versions.

dinooo13 commented:
I'm not sure if the re-request review button worked you may now have no or quite a few requests 🤣

WyriHaximus commented:
> I'm not sure if the re-request review button worked you may now have no or quite a few requests rofl

Looks like even I can't re-request @SimonFrings' review :scream:

WyriHaximus reviewed (approved):
:+1:

SimonFrings commented:
@WyriHaximus Could be a sign, but I am here anyway ^^

@dinooo13 Changes are looking great, the last thing I want to mention is the amount of commits for this. I don't think it is necessary to have 10 commits for this, can you squash them together into one? After that this is ready to be merged! :+1:

dinooo13 commented:
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests

Can u maybe enable this on the repo or am I just not seeing this option? This would make it way easier to merge PRs. But I can also squash manually in the evening I think.

dinooo13 commented:
Everything squashed 😄 :shipit:

SimonFrings reviewed (approved):
Nice one :+1:

SimonFrings commented:
Ping @clue, this is ready for review 👍

clue commented on src/Io/Transaction.php:
```diff
@@ -255,25 +256,38 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
-            ->withoutHeader('Content-Length');
+        $request = $request->withoutHeader('Host');
+
+        if ($statusCode !== 307 || $statusCode !== 308) {
```
Should use status code constants: https://github.com/reactphp/http/pull/432

clue commented on src/Io/Transaction.php:
```diff
@@ -255,25 +256,38 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
-        // naïve approach..
-        $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
+        $body = null;
+        if ($statusCode === 307 || $statusCode === 308) {
```
Should use status code constants: https://github.com/reactphp/http/pull/432

clue replied:
Also, this seems to check the same status codes checked above, perhaps we can refactor this to avoid the duplicate check?

clue commented on src/Io/Transaction.php:
```diff
@@ -255,25 +256,38 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
      * @param UriInterface $location
+     * @param int $statusCode
      * @return RequestInterface
      */
```
Should add a `@throws` annotation here?

clue commented on tests/Io/TransactionTest.php:
```diff
@@ -671,6 +671,107 @@ public function testSomeRequestHeadersShouldBeRemovedWhenRedirecting()
+        // response to the given $request
+        $redirectResponse = new Response(307, array('Location' => 'http://example.com/new'));
+
+        $sender->expects($this->exactly(1))->method('send')->withConsecutive(
```
```suggestion
        $sender->expects($this->once())->method('send')->withConsecutive(
```

clue commented on src/Io/Transaction.php:
```diff
@@ -255,25 +256,38 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
+        }
 
-        return new Request($method, $location, $request->getHeaders());
+        return new Request($method, $location, $request->getHeaders(), $body);
```
Might as well return the request returned from the withers here instead of explicitly creating a new instance, WDYT?

dinooo13 replied:
I refactored the method a bit, so we don't need the new instance 👍🏼

dinooo13 replied:
```
        if (!$this->isTemporaryOrPermanentRedirect($statusCode)) {
            // naïve approach..
            $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
            $request = $request->withMethod($method)->withBody(new EmptyBodyStream());
        } elseif ($request->getBody() instanceof ReadableStreamInterface) {
            throw new \RuntimeException('Unable to redirect request with streaming body');
        }
```

This could now also be refactored to this so we don't have the double if with only the exception but I think this is only cleaner but not really clearer on what it does.

clue commented on src/Io/Transaction.php:
```diff
@@ -12,6 +12,7 @@
 use React\Promise\Deferred;
 use React\Promise\PromiseInterface;
 use React\Stream\ReadableStreamInterface;
+use RuntimeException;
```
Not wrong, but we don't usually add imports for global classes, WDYT?

clue reviewed (commented):
@dinooo13 Some good progress, would love to get this in! :+1: See minor remarks below, otherwise LGTM :+1:

SimonFrings commented on src/Io/Transaction.php:
```diff
@@ -255,25 +258,37 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
-    private function makeRedirectRequest(RequestInterface $request, UriInterface $location)
+    private function makeRedirectRequest(RequestInterface $request, UriInterface $location, $statusCode)
     {
         $originalHost = $request->getUri()->getHost();
```
```suggestion

```
Part A1: I would get red of this line here and add it a few lines below to bring context related code together (**see part A2**)

SimonFrings commented on src/Io/Transaction.php:
```diff
@@ -255,25 +258,37 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
         // Remove authorization if changing hostnames (but not if just changing ports or protocols).
         if ($location->getHost() !== $originalHost) {
             $request = $request->withoutHeader('Authorization');
         }
```
```suggestion
        // Remove authorization if changing hostnames (but not if just changing ports or protocols).
        $originalHost = $request->getUri()->getHost();
        if ($location->getHost() !== $originalHost) {
            $request = $request->withoutHeader('Authorization');
        }
```
Part A2: This is the result from A1.

SimonFrings commented on src/Io/Transaction.php:
```diff
@@ -255,25 +258,37 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
+            $request = $request
+                ->withoutHeader('Content-Type')
+                ->withoutHeader('Content-Length');
+        }
```
```suggestion

```
Part B1: We can also merge this if with the if down below (**see part B2**)

SimonFrings commented on src/Io/Transaction.php:
```diff
@@ -255,25 +258,37 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
+            // naïve approach..
+            $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
+            $request = $request->withMethod($method)->withBody(new EmptyBodyStream());
+        }
```
```suggestion
        if ($statusCode === Response::STATUS_TEMPORARY_REDIRECT || $statusCode === Response::STATUS_PERMANENT_REDIRECT) {
            if ($request->getBody() instanceof ReadableStreamInterface) {
                throw new \RuntimeException('Unable to redirect request with streaming body');
            }
        } else {
            $request = $request
                ->withMethod($request->getMethod() === 'HEAD' ? 'HEAD' : 'GET')
                ->withoutHeader('Content-Type')
                ->withoutHeader('Content-Length')
                ->withBody(new EmptyBodyStream());
        }
```
Part B2: This is the result from B1, I also optimized the code a bit. This way we don't need the extra `isTemporaryOrPermanentRedirect($statusCode)` function. (**See part B3**)

SimonFrings commented on src/Io/Transaction.php:
```diff
@@ -295,4 +310,15 @@ private function progress($name, array $args = array())
+        return
+            $statusCode === Response::STATUS_TEMPORARY_REDIRECT
+            || $statusCode === Response::STATUS_PERMANENT_REDIRECT;
+    }
```
```suggestion

```
Part B3: No need for this function any more, save some lines of code 👍

SimonFrings commented on tests/Io/TransactionTest.php:
```diff
@@ -671,6 +671,107 @@ public function testSomeRequestHeadersShouldBeRemovedWhenRedirecting()
+        $promise = $transaction->send($request);
+
+        $this->setExpectedException('RuntimeException', 'Unable to redirect request with streaming body');
+        \React\Async\await($promise, $loop);
```
```suggestion
        $exception = null;
        $promise->then(null, function ($reason) use (&$exception) {
            $exception = $reason;
        });

        assert($exception instanceof RuntimeException);
        $this->assertEquals('Unable to redirect request with streaming body', $exception->getMessage());
```
To make this part consistent to the other tests we can add this structure.

SimonFrings reviewed (commented):
Hey @dinooo13, I think with refactoring @clue meant to bring related code together. I hope my suggestions aren't too confusing, I couldn't make one big diff out of this (GitHub wouldn't let me 😅)

dinooo13 commented:
LGTM! I applied your suggestions 👍🏼

SimonFrings commented:
@dinooo13 Nice, now you only need to squash your commits ;)

dinooo13 committed:
Preserve method on redirect

dinooo13 commented:
Oh I think I messed something up squashing I will fix this in the evening

dinooo13 commented:
> Oh I think I messed something up squashing I will fix this in the evening

Or not as it seems 😆 thought the checks failed but looks good

clue commented on src/Io/Transaction.php:
```diff
@@ -295,4 +306,5 @@ private function progress($name, array $args = array())
 
         echo PHP_EOL;
     }
+
```
nit: whitespace 🔥

clue commented on tests/Io/TransactionTest.php:
```diff
@@ -674,6 +674,112 @@ public function testSomeRequestHeadersShouldBeRemovedWhenRedirecting()
+            array($this->anything()),
+            array($this->callback(function (RequestInterface $request) use ($that) {
+                $that->assertEquals('POST', $request->getMethod());
+                $that->assertEquals('{"key":"value"}', (string)$request->getBody());
```
LGTM, but perhaps this should also test request headers are the same?

clue commented on tests/Io/TransactionTest.php:
```diff
@@ -674,6 +674,112 @@ public function testSomeRequestHeadersShouldBeRemovedWhenRedirecting()
+        $sender->expects($this->exactly(2))->method('send')->withConsecutive(
+            array($this->anything()),
+            array($this->callback(function (RequestInterface $request) use ($that) {
+                $that->assertEquals('GET', $request->getMethod());
```
LGTM, but perhaps this should also test which request headers are sent?

clue reviewed (commented):
@dinooo13 Thanks for the update! Changes LGTM, but perhaps we can add some additional tests like suggested below?

Also, the documentation currently says "If the original request contained a request body, this request body will never be passed to the redirected request. Accordingly, each redirected request will remove any Content-Length and Content-Type request headers.", should we update this as part of this PR?

Keep it up! :+1:

dinooo13 commented:
Added asserts for the headers in the tests as suggested and 🔥'd the whitespace

SimonFrings commented:
Nice one 👍

Have you also seen @clue's comment regarding the `README.md`?
> Also, the documentation currently says "If the original request contained a request body, this request body will never be passed to the redirected request. Accordingly, each redirected request will remove any Content-Length and Content-Type request headers.", should we update this as part of this PR?

dinooo13 commented:
Ahh I thought this was about checking the headers have been removed in the test with See Other as status code. I will update the Readme!

clue reviewed (approved):
@dinooo13 Thank you very much for looking into this, changes LGTM! Keep it up! :shipit:
